pandas/_libs/groupby.pyx

```diff
@@ -825,27 +828,28 @@ def group_quantile(ndarray[float64_t, ndim=2] out,
     ngroups = len(out)
     counts = np.zeros(ngroups, dtype=np.int64)
     non_na_counts = np.zeros(ngroups, dtype=np.int64)
+    labels_for_lexsort = labels.copy()
+
+    # Put '-1' (NaN) labels as the last group so it does not interfere
+    # with the calculations.
+    if N > 0:
+        na_label_for_sorting = labels.max() + 1
 
     # First figure out the size of every group
     with nogil:
         for i in range(N):
             lab = labels[i]
-            if lab == -1:  # NA group label
+            if lab == -1:
+                labels_for_lexsort[i] = na_label_for_sorting
                 continue
 
             counts[lab] += 1
             if not mask[i]:
                 non_na_counts[lab] += 1
 
-    # Get an index of values sorted by labels and then values
-    if labels.any():
-        # Put '-1' (NaN) labels as the last group so it does not interfere
-        # with the calculations.
-        labels_for_lexsort = np.where(labels == -1, labels.max() + 1, labels)
-    else:
-        labels_for_lexsort = labels
+    # Get an index of values sorted by values and then labels
     order = (values, labels_for_lexsort)
-    sort_arr = np.lexsort(order).astype(np.int64, copy=False)
```

Comment:
FWIW the big perf gain comes from moving this call out of cython and not repeating it for each column

Reply:
Ok, makes sense. If you're planning on doing that, I'll just close this for now and collect the other cleanups elsewhere

Reply:
I was actually suggesting you do that in this PR

Reply:
Ah sounds good :) I'll take a look at that